### PR TITLE
Fixes for enabling pg_cron extension

### PR DESCRIPTION
**Pull request by markdboyd**
Related to #270 

## Changes proposed in this pull request:

- Addresses PR feedback from #276
- Adds unit tests

## Security considerations

None

### CODE REVIEW

File: services/rds/parameter_group.go [hunk 1]
```diff
@@ -13,25 +13,75 @@ import (
...
-		d *dedicatedDBAdapter,
-		svc rdsiface.RDSAPI,
-	) (string, error)
+	ProvisionCustomParameterGroupIfNecessary(i *RDSInstance) error
+}
+
+type parameterGroupAdapter struct {
+	svc                  rdsiface.RDSAPI
+	settings             config.Settings
+	parameterGroupPrefix string `default:"cg-aws-broker-"`
```
Comment (jameshochadel) [inline services/rds/parameter_group.go:25]: I've never seen the `default` tag before, but it's a neat idea! Where is it used? I don't see any calls to `reflect` in your code; does a package you're importing support it?

File: services/rds/parameter_group.go [hunk 2]
```diff
@@ -13,25 +13,75 @@ import (
...
-	provisionCustomParameterGroupIfNecessary(
-		i *RDSInstance,
-		d *dedicatedDBAdapter,
-		svc rdsiface.RDSAPI,
-	) (string, error)
+	ProvisionCustomParameterGroupIfNecessary(i *RDSInstance) error
+}
+
+type parameterGroupAdapter struct {
+	svc                  rdsiface.RDSAPI
```
Comment (jameshochadel) [inline services/rds/parameter_group.go:23]: Love this refactor. May I suggest renaming this to `rds`? That way your calls to it will read as `rds.DoSomething()`, which reads nicely.

File: services/rds/parameter_group.go [hunk 3]
```diff
@@ -13,25 +13,75 @@ import (
...
+	if err != nil {
+		log.Println(err.Error())
+		return fmt.Errorf("encountered error applying parameter group: %w", err)
+	}
+	i.ParameterGroupName = parameterGroupName
+	return nil
 }
 
-type parameterGroupAdapter struct{}
+// search out all the parameter groups that we created and try to clean them up
```
Comment (jameshochadel) [inline services/rds/parameter_group.go:47]: I think Go doc comments require the first word to be the function name and are written by convention as full sentences, e.g. "CleanupCustomParameterGroups searches out all the parameter groups that we created and tries to clean them up." (It looks nice on pkg.go.dev.)

File: services/rds/parameter_group.go [hunk 4]
```diff
@@ -13,25 +13,75 @@ import (
...
+func (p *parameterGroupAdapter) CleanupCustomParameterGroups() {
+	input := &rds.DescribeDBParameterGroupsInput{}
+	err := p.svc.DescribeDBParameterGroupsPages(input,
+		func(pgroups *rds.DescribeDBParameterGroupsOutput, lastPage bool) bool {
```
Comment (jameshochadel) [inline services/rds/parameter_group.go:51]: Edit: For this and other comments on this method, I realized after reading further that you're not the author, you just moved it. But, if you want to do some extra cleanup, have at it!

Go programmers often don't mind long-ish lines, but try to avoid deeply indented code; I'd move this signature up a line to save yourself a level of indentation. (You'll notice some Go idioms, like `if err  { return err }` with no `else` clause are designed to prevent deep nesting of blocks.)

File: services/rds/parameter_group.go [hunk 5]
```diff
@@ -13,25 +13,75 @@ import (
...
+			// If the pgroup matches the prefix, then try to delete it.
+			// If it's in use, it will fail, so ignore that.
+			for _, pgroup := range pgroups.DBParameterGroups {
+				matched, err := regexp.Match("^"+p.parameterGroupPrefix, []byte(*pgroup.DBParameterGroupName))
+				if err != nil {
+
```
Comment (jameshochadel) [inline services/rds/parameter_group.go:57]: Spare line :)

File: services/rds/parameter_group.go [hunk 6]
```diff
@@ -13,25 +13,75 @@ import (
...
+						// and we should log that.
+						log.Printf("There was an error cleaning up the %s parameter group.  The error was: %s", *pgroup.DBParameterGroupName, err.Error())
+					}
+				}
+			}
+			return true
+		})
+	if err != nil {
+		log.Printf("Could not retrieve list of parameter groups while cleaning up: %s", err.Error())
+		return
```
Comment (jameshochadel) [inline services/rds/parameter_group.go:80]: This can be removed, since the method returns after the if statement anyway.

**Review by jameshochadel — COMMENTED** [review]
Nicely done. I left a few comments after going over it with a fine-toothed comb. This is a great improvement!

File: services/rds/parameter_group.go (hunk 1 above)
Comment (markdboyd) [reply services/rds/parameter_group.go:25]: I might be dense, but I thought `default` was simply a way to set a default value for `parameterGroupPrefix` in the struct. So that if you initialize `&parameterGroupAdapter`, the value of `parameterGroupPrefix` by default would be `cg-aws-broker-`. But I suppose I should write a test to verify

File: services/rds/parameter_group.go (hunk 6 above)
Comment (markdboyd) [reply services/rds/parameter_group.go]: 030848f003e02ab1a07099eeb1e1848355c15314

File: services/rds/parameter_group.go (hunk 5 above)
Comment (markdboyd) [reply services/rds/parameter_group.go]: 030848f003e02ab1a07099eeb1e1848355c15314

File: services/rds/parameter_group.go (hunk 3 above)
Comment (markdboyd) [reply services/rds/parameter_group.go]: 030848f003e02ab1a07099eeb1e1848355c15314

File: services/rds/parameter_group.go (hunk 2 above)
Comment (markdboyd) [reply services/rds/parameter_group.go]: 030848f003e02ab1a07099eeb1e1848355c15314

File: services/rds/parameter_group.go (hunk 1 above)
Comment (markdboyd) [reply services/rds/parameter_group.go]: Yeah this blatantly didn't work. I've refactored to use a constructor: 92c9234514dac553e62e66cca4a3fb9758c9d131

Comment (jameshochadel) [reply services/rds/parameter_group.go]: A little Go background: The third token on a struct field, after the field name and type, is called a struct tag. There's a good crash course on the `json` tag here: https://www.digitalocean.com/community/tutorials/how-to-use-struct-tags-in-go 

And an explanation of how it's actually read, using `reflect`, in the top answer here: https://stackoverflow.com/questions/10858787/what-are-the-uses-for-struct-tags-in-go

The key thing is that struct tags don't have any behavior by default; a package (yours or another one) has to read them. So if nothing is reading `default`, it won't do anything.

File: services/rds/parameter_group.go [hunk 7]
```diff
@@ -13,25 +13,85 @@ import (
...
+	settings             config.Settings
+	parameterGroupPrefix string
+}
+
+type paramDetails struct {
+	value       string
+	applyMethod string
+}
+
+func NewParameterGroupAdapater(rds rdsiface.RDSAPI, settings config.Settings) *parameterGroupAdapter {
```
Comment (jameshochadel) [inline services/rds/parameter_group.go:33]:
```suggestion
func NewParameterGroupAdapter(rds rdsiface.RDSAPI, settings config.Settings) *parameterGroupAdapter {
```

Last change from me 😄

Comment (markdboyd) [reply services/rds/parameter_group.go]: 2081aceacc39f3de08b7ac3a92eacd06ab9a51d9

**Review by jameshochadel — APPROVED** [review]

File: services/rds/parameter_group.go [hunk 8]
```diff
@@ -137,40 +140,49 @@ func (p *parameterGroupAdapter) checkIfParameterGroupExists(pgroupName string) b
...
+}
+
+func (p *parameterGroupAdapter) setParameterGroupName(i *RDSInstance) {
+	if i.ParameterGroupName != "" {
+		return
+	}
+	i.ParameterGroupName = p.getParameterGroupName(i)
+}
+
 // This function will return the a custom parameter group with whatever custom
```
Comment (jameshochadel) [inline services/rds/parameter_group.go:161]: Docstring needs updated to reflect new signature

File: services/rds/parameter_group.go [hunk 9]
```diff
@@ -207,58 +219,106 @@ func (p *parameterGroupAdapter) needCustomParameters(i *RDSInstance) bool {
...
-		describeEngDefaultParamsInput.Marker = result.EngineDefaults.Marker
-		more = describeEngDefaultParamsInput.Marker != nil && *describeEngDefaultParamsInput.Marker != ""
-	}
-	return "", nil
+		return foundValue
+	})
+	return err
 }
 
-func (p *parameterGroupAdapter) buildCustomSharePreloadLibrariesParam(
```
Comment (jameshochadel) [inline services/rds/parameter_group.go]: Could you split this refactor into a separate PR?

File: services/rds/broker.go [hunk 10]
```diff
@@ -239,6 +219,44 @@ func (broker *rdsBroker) ModifyInstance(c *catalog.Catalog, id string, modifyReq
...
+			http.StatusBadRequest,
+			"Cannot enable and disable pg_cron extension at the same time",
+		)
+	}
+	existingInstance.DisablePgCron = options.DisablePgCron
+	existingInstance.EnablePgCron = options.EnablePgCron
+	return nil
+}
+
+func (broker *rdsBroker) ModifyInstance(c *catalog.Catalog, id string, modifyRequest request.Request, baseInstance base.Instance) response.Response {
```
[the remark this answers is not in the document]
Comment (markdboyd) [reply services/rds/broker.go:233]: I didn't move this method, I just moved the logic for parsing a modify request into options into `ParseModifyOptions` to make it easier to test

File: services/rds/parameter_group.go [hunk 11]
```diff
@@ -137,40 +140,49 @@ func (p *parameterGroupAdapter) checkIfParameterGroupExists(pgroupName string) b
...
+	// so the parameter group name should remain consistent
+	return p.parameterGroupPrefix + i.FormatDBName()
+}
+
+func (p *parameterGroupAdapter) setParameterGroupName(i *RDSInstance) {
+	if i.ParameterGroupName != "" {
+		return
+	}
+	i.ParameterGroupName = p.getParameterGroupName(i)
+}
```
[the remark this answers is not in the document]
Comment (markdboyd) [reply services/rds/parameter_group.go:159]: This makes sense

File: main_test.go [hunk 12]
```diff
@@ -446,7 +446,7 @@ func TestCreateRDSPostgreSQLWithEnablePgCron(t *testing.T) {
 		t.Error("The instance should have metadata")
 	}
 
-	if !i.EnablePgCron {
+	if !*i.EnablePgCron {
```
Comment (jameshochadel) [inline main_test.go:449]: This will cause a segmentation violation if the pointer is nil, because you can't do boolean operations like `!` on `nil`. You'll have to check for nil first. (Same below.)

Example: https://go.dev/play/p/_V_0ZgMCxTP

**Review by jameshochadel — APPROVED** [review]
